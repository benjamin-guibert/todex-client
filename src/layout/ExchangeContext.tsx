import { createContext, useCallback, useEffect, useRef, useState } from 'react'
import { Web3Provider } from '@ethersproject/providers'
import { Token } from 'contracts/Token'
import { ExchangeHandler, getEthBalance, getTokenBalance, initializeContract } from 'libraries/contracts/exchange'

export interface ExchangeContextValue {
  initialized: boolean
  ethBalance: string | undefined
  tokenBalance: string | undefined
  initialize: (provider: Web3Provider, token: Token) => Promise<boolean>
  setAccount: (account: string | undefined) => void
  updateBalances: () => Promise<void>
}

export const useExchangeContext = (): ExchangeContextValue => {
  const exchangeHandlerRef = useRef<ExchangeHandler>()
  const [initialized, setInitialized] = useState<boolean>(false)
  const [ethBalance, setEthBalance] = useState<string | undefined>()
  const [account, setAccount] = useState<string | undefined>()
  const [tokenBalance, setTokenBalance] = useState<string | undefined>()

  const initializeValue = async (provider: Web3Provider, token: Token) => {
    exchangeHandlerRef.current = initializeContract(provider, token)
    setInitialized(!!exchangeHandlerRef.current)

    return !!exchangeHandlerRef.current
  }

  const updateBalancesValue = useCallback(async () => {
    if (!exchangeHandlerRef.current || !account) {
      return
    }
    setEthBalance((await getEthBalance(exchangeHandlerRef.current, account))?.toString())
    setTokenBalance((await getTokenBalance(exchangeHandlerRef.current, account))?.toString())
  }, [account])

  const setAccountValue = (account: string | undefined) => {
    setAccount(account)
  }

  useEffect(() => {
    updateBalancesValue()
  }, [updateBalancesValue, account])

  return {
    initialized,
    ethBalance,
    tokenBalance,
    initialize: initializeValue,
    setAccount: setAccountValue,
    updateBalances: updateBalancesValue,
  }
}

export const ExchangeContext = createContext<ExchangeContextValue>({
  initialized: false,
  ethBalance: undefined,
  tokenBalance: undefined,
  initialize: async () => false,
  setAccount: () => undefined,
  updateBalances: async () => undefined,
})
