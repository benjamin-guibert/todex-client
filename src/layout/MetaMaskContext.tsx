import { createContext, useRef, useState } from 'react'
import { Web3Provider } from '@ethersproject/providers'
import { MetaMaskHandler, initialize, connect, getAccount, getBalance } from 'libraries/metamask'

export enum InitializationStatus {
  None = 0,
  Initializing = 1,
  Initialized = 2,
  NotInstalled = 3,
  Error = 4,
}

export interface MetaMaskContextValue {
  initializationStatus: InitializationStatus
  account: string | undefined
  balance: string | undefined
  initialize: () => Promise<Web3Provider | undefined>
  connect: () => Promise<string | undefined>
  getAccount: () => Promise<string | undefined>
  getBalance: (account: string) => Promise<string | undefined>
}

export const useMetaMaskContext = (): MetaMaskContextValue => {
  const metaMaskHandlerRef = useRef<MetaMaskHandler>()
  const [initializationStatus, setInitializationStatus] = useState<InitializationStatus>(InitializationStatus.None)
  const [account, setAccount] = useState<string | undefined>()
  const [balance, setBalance] = useState<string | undefined>()

  const initializeValue = async () => {
    try {
      metaMaskHandlerRef.current = await initialize()
      if (!metaMaskHandlerRef.current) {
        setInitializationStatus(InitializationStatus.NotInstalled)
      }

      setInitializationStatus(InitializationStatus.Initialized)

      getAccountValue()
    } catch {
      setInitializationStatus(InitializationStatus.Error)
    }

    return metaMaskHandlerRef.current?.provider
  }

  const connectValue = async () => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const account = await connect(metaMaskHandlerRef.current)
    setAccount(account)

    return account
  }

  const getAccountValue = async () => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const account = await getAccount(metaMaskHandlerRef.current)
    setAccount(account)

    return account
  }

  const getBalanceValue = async (account: string) => {
    if (!metaMaskHandlerRef.current) {
      return undefined
    }

    const balance = (await getBalance(metaMaskHandlerRef.current, account))?.toString()

    setBalance(balance)

    return balance
  }

  return {
    initializationStatus,
    account,
    balance,
    initialize: initializeValue,
    connect: connectValue,
    getAccount: getAccountValue,
    getBalance: getBalanceValue,
  }
}

export const MetaMaskContext = createContext<MetaMaskContextValue>({
  initializationStatus: InitializationStatus.None,
  account: undefined,
  balance: undefined,
  initialize: async () => undefined,
  connect: async () => undefined,
  getAccount: async () => undefined,
  getBalance: async () => undefined,
})
