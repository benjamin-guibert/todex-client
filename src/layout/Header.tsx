import React, { FC, useContext } from 'react'
import { InitializationStatus, MetaMaskContext } from './MetaMaskContext'
import { ethers } from 'ethers'
import { printAmount, shortenAddress } from 'libraries/helpers'
import { ExchangeContext } from './ExchangeContext'
import Container from 'react-bootstrap/Container'
import Navbar from 'react-bootstrap/Navbar'
import Button from 'react-bootstrap/Button'
import StyledHeader from './Header.style'

const printBalance = (balance?: string) => {
  return printAmount(balance || null)
}

const Header: FC = () => {
  const { initializationStatus, account, connect } = useContext(MetaMaskContext)
  const { ethBalance, tokenBalance } = useContext(ExchangeContext)

  return (
    <StyledHeader bg="dark" variant="dark">
      <Container fluid>
        <Navbar.Brand>ToDEX</Navbar.Brand>
        {initializationStatus == InitializationStatus.Initialized && (
          <>
            {!!ethBalance && <Navbar.Text>{`${ethers.constants.EtherSymbol} ${printBalance(ethBalance)}`}</Navbar.Text>}
            {!!tokenBalance && <Navbar.Text>{`ToDEX ${printBalance(tokenBalance)}`}</Navbar.Text>}
            {!!account && <Navbar.Text>{shortenAddress(account)}</Navbar.Text>}
            {!account && <Button onClick={connect}>Connect Wallet</Button>}
          </>
        )}
      </Container>
    </StyledHeader>
  )
}

export default Header
