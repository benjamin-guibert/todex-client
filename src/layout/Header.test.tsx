import * as React from 'react'
import renderer, { ReactTestRenderer } from 'react-test-renderer'
import { InitializationStatus, MetaMaskContext, MetaMaskContextValue } from './MetaMaskContext'
import { ExchangeContext, ExchangeContextValue } from './ExchangeContext'
import Button from 'react-bootstrap/Button'
import Header from './Header'

const ACCOUNT_ADDRESS = '0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266'

const createMetaMaskContextValue = (): MetaMaskContextValue => {
  return {
    account: undefined,
    balance: undefined,
    initializationStatus: InitializationStatus.None,
    initialize: jest.fn(),
    connect: jest.fn(),
    getAccount: jest.fn(),
    getBalance: jest.fn(),
  }
}

const createExchangeContextValue = (): ExchangeContextValue => {
  return {
    initialized: true,
    ethBalance: undefined,
    tokenBalance: undefined,
    trades: [],
    initialize: jest.fn(),
    setAccount: jest.fn(),
    updateBalances: jest.fn(),
  }
}

const renderComponent = (metaMaskContextValue: MetaMaskContextValue, exchangeContextValue: ExchangeContextValue) =>
  renderer.create(
    <MetaMaskContext.Provider value={metaMaskContextValue}>
      <ExchangeContext.Provider value={exchangeContextValue}>
        <Header />
      </ExchangeContext.Provider>
    </MetaMaskContext.Provider>
  )

describe('<Header />', () => {
  let component: ReactTestRenderer

  it('should render when MetaMask not initialized', () => {
    const metaMaskContextValue = createMetaMaskContextValue()
    const exchangeContextValue = createExchangeContextValue()

    component = renderComponent(metaMaskContextValue, exchangeContextValue)

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render when MetaMask initialized', () => {
    const metaMaskContextValue = {
      ...createMetaMaskContextValue(),
      initializationStatus: InitializationStatus.Initialized,
    }
    const exchangeContextValue = {
      ...createExchangeContextValue(),
      account: ACCOUNT_ADDRESS,
      ethBalance: '10500000000000000000',
      tokenBalance: '1000500000000000000000',
    }

    component = renderComponent(metaMaskContextValue, exchangeContextValue)

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should render when no account connected', () => {
    const metaMaskContextValue = {
      ...createMetaMaskContextValue(),
      initializationStatus: InitializationStatus.Initialized,
    }
    const exchangeContextValue = createExchangeContextValue()

    component = renderComponent(metaMaskContextValue, exchangeContextValue)

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should connect on click', () => {
    const metaMaskContextValue = {
      ...createMetaMaskContextValue(),
      initializationStatus: InitializationStatus.Initialized,
    }
    const exchangeContextValue = {
      ...createExchangeContextValue(),
      account: ACCOUNT_ADDRESS,
      ethBalance: '10500000000000000000',
      tokenBalance: '1000500000000000000000',
    }
    component = renderComponent(metaMaskContextValue, exchangeContextValue)

    component.root.findByType(Button).props.onClick()

    expect(metaMaskContextValue.connect).toBeCalledTimes(1)
  })
})
