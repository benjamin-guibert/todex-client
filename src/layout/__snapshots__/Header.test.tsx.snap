// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render when MetaMask initialized 1`] = `
<nav
  className="sc-bdvvtL dqBChi navbar navbar-expand navbar-light bg-light"
>
  <div
    className="container-fluid"
  >
    <span
      className="navbar-brand"
    >
      ToDEX
    </span>
    <span
      className="navbar-text"
    >
      Îž 10.5
    </span>
    <span
      className="navbar-text"
    >
      TDX 1,000.5
    </span>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      Connect Wallet
    </button>
  </div>
</nav>
`;

exports[`<Header /> should render when MetaMask not initialized 1`] = `
<nav
  className="sc-bdvvtL dqBChi navbar navbar-expand navbar-light bg-light"
>
  <div
    className="container-fluid"
  >
    <span
      className="navbar-brand"
    >
      ToDEX
    </span>
  </div>
</nav>
`;

exports[`<Header /> should render when no account connected 1`] = `
<nav
  className="sc-bdvvtL dqBChi navbar navbar-expand navbar-light bg-light"
>
  <div
    className="container-fluid"
  >
    <span
      className="navbar-brand"
    >
      ToDEX
    </span>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[MockFunction]}
      type="button"
    >
      Connect Wallet
    </button>
  </div>
</nav>
`;
